Add `depth` as an additional variable in Sv dataset

leewujung opened the pull request:
This addresses #578.

The current prototype defaults to assuming that `water_level` exists in the dataset and that the transducers are mounted vertically, pointing downward, like in a ship-based survey.

Users have the option to set `water_level` and `tilt` via input directly, since many times these data are not in the data files generated by the instruments.

@valentina-s @emiliom : I'd appreciate comments on whether we should replace the original `echo_range` variable or keep `depth`. These 2 variables are of the same dimension as the echo data (e.g., `backscatter_r`) so it would blow up the size of the dataset -- maybe we should replace it?  Either way we'd need to be careful on the provenance to record the operations.

**Updates:**
- ~Note: I am keeping this as a draft for now because it builds on top of #735 and somehow the CI is failing repeated not due to fault of the tests but some bucket creation problems.~
- Due to the addition of the `consolidate` subpackage, this PR also includes:
  - removing a redundant `swap_dims_channel_frequency` previously in `preprocess`
  - moving `add_depth` and `swap_dims_channel_frequency` to the `consolidate` subpackage

### Tasks
- [x] add function
- [x] add tests

leewujung commented:
One thing that I think can be expanded on this function is to allow `tilt` (or ds["tilt"] ) to be a DataArray instead of a scalar, but I think we can add that expansion in a separate PR and get this simplistic but useful function in first.

emiliom commented on echopype/consolidate/api.py:
```diff
@@ -7,6 +7,111 @@
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Dataset for which the dimension will be swapped.
```
This definition likely stems from an earlier version of what `add_depth` would do. Currently it doesn't swap dimensions
```suggestion
        Source Sv dataset to which a depth variable will be added.
```

emiliom commented on echopype/consolidate/api.py:
```diff
@@ -7,6 +7,111 @@
+        Whether or not the transducers point downward.
+        Default to True.
+    tilt : float
+        Tilt angle (degree) if transducers are not mounted vertically (when `vertical=False`).
+        Required if `vertical=False` and `tilt` does not exist in the dataset.
```
As we know, I'm less of a fan of redundant descriptions than you are :laughing: . So, here's a leaner description, for your consideration:
```suggestion
        Tilt angle (degree). Required if transducers are not mounted vertically
        (`vertical=False`) and `tilt` does not exist in the dataset.
```

leewujung replied:
Haha, sure!

emiliom commented on echopype/consolidate/api.py:
```diff
@@ -7,6 +7,111 @@
+        mult = -1  # flip upside down (closer to transducer is deeper)
+
+    # Compute depth
+    ds["depth"] = mult * ds["echo_range"] * np.cos(tilt / 180 * np.pi) + water_level
```
Trivial tweak. I prefer to avoid multiplications after a denominator, for clarity.
```suggestion
    ds["depth"] = mult * ds["echo_range"] * np.cos(np.pi * tilt / 180) + water_level
```

emiliom replied:
Also, some of the inherited attributes of the new depth variable should be changed:
- Change `long_name` to, say, "Depth"
- Add a `standard_name` with value "depth"

leewujung replied:
> I prefer to avoid multiplications after a denominator, for clarity.

I can see the advantage, but here because the conversion is from degree to radians, breaking the sequence seems weird.

leewujung replied:
attributes added!

emiliom commented:
Regarding the interpretation of `water_level` and the depth calculation `ds["depth"] = mult * ds["echo_range"] * np.cos(tilt / 180 * np.pi) + water_level`: I believe (but haven't looked it up in the SONAR-netcdf4 manual) that we learned that your likely interpretation of `water_level` (depth below the water surface, right?) is not quite what's in the standard and nominally in echopype since 0.6.0. In the standard, `water_level` is the "z-axis distance from the platform coordinate system origin to the sonar transducer", and it requires `vertical_offset` ("Platform vertical offset from nominal") in order to calculate the actual depth below water surface.

In practice, if I remember correctly from our 0.6.0 development, `vertical_offset` is often not available and all that's available is a `water_level` variable that may or may not be heave. So, in practice, let's not worry about it for now.

A more important issue is that since 0.6.0, `water_level` is *not* a scalar. It's a vector with dimensions that don't match the `echo_range` dimensions. I *think* this vector is typically a constant, propagated value. So, an easy and cheap approach for this initial implementation could be to take the first `water_level` value and use it as a scalar.

emiliom commented:
Regarding `tilt`, that's a variable that's not found in the SONAR-netcdf4 v1 convention. I remember us discussing it, and I thought I had seen it in one of the OOI moored echosounders (the AZFP's, right?), but it's not in the EK60 example in https://osoceanacoustics.github.io/echopype-examples/ms_OOI_EK60_mooringtimeseries.html. Can tilt be calculated from other Platform orientation variables found in the standard, such that it could be calculated for that OOI EK60 example?

Assuming it *can* be calculated, we'd have the same issue as with `water_level`, of having to convert from a vector to a single scalar value. Again, things that could be deferred to the next release.

emiliom commented on echopype/tests/consolidate/test_consolidate.py:
```diff
@@ -1,6 +1,162 @@
+    except ValueError:
+        ...
+
+    # user input water_level
```
Just to be more explicit, maybe add a comment:
```
# Use defaults `vertical=True` and `downward=True`

leewujung replied:
I removed the option `vertical`

emiliom commented:
I'm done reviewing this. I also ran `test_consolidate.py` locally, and all was fine. `add_depth` is really simple! That's great.

See my various comments. Most of them are minor or academic comments, except the ones about the expectation that the `water_level` dataarray (and `tilt`, but that's a rare case) found in the Sv dataset is a scalar. Your `test_add_depth` tests don't currently include a case with an Sv dataset that does have a `water_level` variable (as generated by `open_raw`), so that case is not covered.

leewujung commented:
Alright, I made the following updates:
- simplified the function to only allow scalar inputs
- added notes to explain future expansion
- removed the unnecessary boolean option `vertical` -- since `tilt` says it all on whether or not something is vertical or not!

leewujung commented:
I'll self-merge and get things ready for release.
